zulip/zulip
# Add per-domain settings for whether invitations are required

https://github.com/zulip/zulip/issues/14469
discussed in https://chat.zulip.org/#narrow/stream/92-learning/topic/On.20boarding.20mechanics
**Testing Plan:** 
manually tested for the following:
case1:
`email_restricted_to_domains=True`
`default_invite_required = True or False`
* domains with `invite_required = False` will be allowed to join without invite.
* Others in allowed domains with `invite_required = True` will only be allowed to join with an invite
* all others not in allowed domains will be restricted from joining no matter how they try to join.

case2:
`email_restricted_to_domains=False`
`default_invite_required = True`
* domains with `invite_required = False` will be allowed to join without invite.
* All other if in allowed domains(with `invite_required = True`) or not will be allowed to join only through an invite.

case3:
`email_restricted_to_domains=False`
`default_invite_required = False`
* Only domains with `invite_required = True` require invite to join the realm.

I think the above covers all cases, all requirements to close the issue. If I  miss something I'd be happy to implement that.

Didn't write automated tests. Will write once it's confirmed the code changes are right and is how it's expected to be.
